### ISSUE TITLE
help with asymmetric inversion

### ISSUE THREAD

**Issue by lemmatum** [PyAbel/PyAbel#236]
Hi,

I'm trying to invert some laser-plasma interferometric data where the only symmetry I have is cylindrical (along the laser axis). 

Here is what the data looks like (the laser axis is vertical along the 0 pixel axis):
![quadrant](https://user-images.githubusercontent.com/28945309/48102930-79fcaf00-e1ea-11e8-97fa-8321bf5318c6.png)

Here is what it would like mirror symmetrized about the laser axis:
![symmetrized](https://user-images.githubusercontent.com/28945309/48102935-7ec16300-e1ea-11e8-9d67-5e63ed1b4965.png)

I've been trying to use the basex method to Abel invert the data (I've tried using it on both the mirrored and non-mirrored images), but it yields odd results. I would expect more of the density to end up at the edge of the image where there is a shock, but instead it seems to smooth things out (as though it were assuming spherical symmetry or an incorrect center axis).

Here is the code snippet I am using:
```Python
invertedImg = abel.Transform(transImg, 
                             direction='inverse', 
                             method='basex',
                             symmetry_axis=0,
#                             center=center,
                             ).transform
```
And this is the result of applying the transform on the mirror symmetrized image:

![symmetrized_inverted](https://user-images.githubusercontent.com/28945309/48103137-2b034980-e1eb-11e8-9be0-8d7fd2a5415e.png)

I mean, it even yields negative values, which cannot be the case.... Am I doing something wrong here?

**Comment by DanHickstein** [comment]
Hi Pawel!

Thanks for pointing this out. The `abel.Transform()` function is intended to operate on left-right symmetric images. If you want to transform just the right-side array, then you can call the individual transform functions, for example, `abel.basex,basex_transform`.

When you symmetrized the image and passed it to `abel.Transform`, you should have gotten a sensible result. In your last image, I can see a faint ring corresponding to your data. I wonder if you re-scaled the clims of your plot to just the positive values -- perhaps `clim=(0,1.0e15)` -- if you would see your data transformed.

While not physical for most experiments, negative values are mathematically possible in the Abel transform, and they can occur in PyAbel, especially if your data is noisy. Also, most transform methods accumulate noise near the centerline. Often, these values are fairly extreme, and require that you manually scale the colormap.

**Comment by stggh** [comment]
This is my attempt.
`basex` inverse Abel transform:
![plasabel](https://user-images.githubusercontent.com/10932229/48104163-525b2480-e286-11e8-9a99-fbdfe891ea67.png)
Slice, the last image row:
![plasabelrow](https://user-images.githubusercontent.com/10932229/48104165-5424e800-e286-11e8-85af-7bd43846dd6e.png)

```python
In [1]: import numpy as np

In [2]: import abel

In [3]: import matplotlib.pyplot as plt

In [4]: im = np.loadtxt('plasma.txt')

In [5]: aim = abel.basex.basex_transform(im)
Loading basis sets...
Updating regularization...

In [6]: plt.imshow(aim, vmin=0, vmax=1)
Out[6]: <matplotlib.image.AxesImage at 0x7fda2b3a6da0>

In [7]: plt.savefig('plasAbel.png', dpi=75)

In [8]: plt.show()

In [9]: plt.plot(aim[-1])
Out[9]: [<matplotlib.lines.Line2D at 0x7fda040e6828>]

In [10]: plt.savefig('plasAbelrow.png', dpi=75)

In [11]: plt.show()
```

**Comment by lemmatum** [comment]
Alright, so I'm not doing things completely wrong! 😝 

I took a normalized lineout (horizontally at the 100 pixel index) of the symmetrized image before and after inversion and this is what I get:
![lineout_100px](https://user-images.githubusercontent.com/28945309/48143464-11561680-e26c-11e8-9bb7-6dd3134891b1.png)

It looks like there is indeed noise accumulating at the center as well as at the edge. Is there some filtering that I need to do or should I use a method other than basex to get physically realistic values?

Here is another lineout for pixel index 25, which has some extreme noise at the center line:
![lineout_25px](https://user-images.githubusercontent.com/28945309/48143660-83c6f680-e26c-11e8-811c-89c74cb429f1.png)


And here is the data in case you want to tinker:
[croppedData.txt](https://github.com/PyAbel/PyAbel/files/2558188/croppedData.txt)

**Comment by lemmatum** [comment]
Looks like Abel inversion has an additional distance unit in the denominator compared to the input values. So we have

```[cm**-3 / pixels]``` and we want ```[cm**-3]```

so we do

```[cm**-3 / pixels] * [pixels / cm] * [cm]```

where ```[pixels / cm]``` is just a spatial conversion factor for the image
and ```[cm]``` is the average length scale of the plasma initially used to convert from line averaged phase shift to density.

This now makes the units and magnitudes look right!
![symmetrized_inverted_scaled](https://user-images.githubusercontent.com/28945309/48151232-f8eef780-e27d-11e8-9787-d70a6b6ee4b5.png)


Did I miss any other factors?

**Comment by DanHickstein** [comment]
Looks like it's working now!

> I also applied a Gaussian filter to smooth things out, and that seems to improve things substantially.

Yeah, I think that most Abel transform algorithms have trouble with noisy images. Smoothing first can improve the transform results in many situations.

Regarding the units, you can also pass the `dr=` keyword to the transform, which is the distance of each pixel. So, `[cm/pixel]` in your terminology. Otherwise, the transform just assumes that `dr=1`, which, in your case, is basically assuming that each pixel is 1 cm.

Since you are using basex, you may want to use the "regularization parameter" to suppress the centerline noise and to provide additional smoothing. Try setting `reg=100` and keep increasing it until you see an effect. You may be able to avoid the gaussian smoothing and just use the regularization parameter.

**Comment by lemmatum** [comment]
@DanHickstein 
Good to know about the `dr=` parameter!
I don't see a `reg` parameter in the docs and I get errors when I try to use it as an input. Perhaps this feature was removed?

**Comment by DanHickstein** [comment]
The `reg` parameter feature was added very recently. You will need to install the most recent version of PyAbel from this github repo in order to use it. The readthedocs documentation has a brief description: https://pyabel.readthedocs.io/en/latest/abel.html#module-abel.basex

**Comment by lemmatum** [comment]
I'm on v0.8.0, and yet no reg parameter... how recently was this added?

**Comment by DanHickstein** [comment]
It was added about a month ago, in PR #227, I believe. We haven't made a release since it was added, so you'll need to download this repo and install with `python setup.py install`

**Comment by stggh** [comment]
> noise accumulating at the center as well as at the edge

The intensity upturn at the edge is due to the non-zero background  `croppedData.min() = 5.5e+16`, subtract this background before the inverse Abel transform.

IMO prefiltering input data is bad. The `croppedData.txt` has oscillations associated with the spine interpolation. Improve the signal/noise in your experiment ;-)

**Comment by DanHickstein** [comment]
> IMO prefiltering input data is bad.

We can't all have perfect data Steve :). A little gaussian blur never hurt anyone. Although I think that the `reg` parameter in basex would really help in this case.

**Comment by stggh** [comment]
`reg` does help suppress the centre-line noise, but the underlying structure simply mirrors the input image data. 

![reg](https://user-images.githubusercontent.com/10932229/48164234-0667b880-e336-11e8-802d-9d9bfe611369.png)

**Comment by lemmatum** [comment]
@stggh as mentioned in a previous comment, I removed the background and it didn't seem to do much... Also that "background" is likely real signal, it's just lower density gas.

>IMO prefiltering input data is bad. The croppedData.txt has oscillations associated with the spine interpolation. Improve the signal/noise in your experiment ;-)

I didn't use a spline, I used a Gaussian blur which is the same as reducing the resolution of one's instrument!

>reg does help suppress the centre-line noise, but the underlying structure simply mirrors the input image data.

Yea I wouldn't be surprised if there was some dense plasma along the center, considering the laser is punching through a rod of solid material which causes a blast wave to expand into a background gas.

**Comment by stggh** [comment]
> removed the background

Sorry, I missed that point.  My transform, above, heads below zero, which does indicate too much background subtracted. 

Note, in general the signal onset follows the instrument function and so a sudden edge step is unphysical.

Also, I am not meaning to pan your nice experiments, I play in very a different experimental world.

**Comment by lemmatum** [comment]
> Note, in general the signal onset follows the instrument function and so a sudden edge step is unphysical.
So would a Gaussian be too "sharp" then?

> Also, I am not meaning to pan your nice experiments, I play in very a different experimental world.
No worries - us plasma physicists are used to low signal-to-noise, small sample size, and large error bars 😛

**Comment by lemmatum** [comment]
>  the edge data must be zero (in the real world)

Not necessarily. That "edge" is the blast wave and it is expanding into a background gas that has been at least partially ionized. So that low density region may be really non-zero.
Now whether the Abel inversion requires a curve that falls to zero or not is another matter...

> I see nothing wrong with the unblurred data transform in my first post:
Yea I have to play around a bit more with blurred vs unblurred. The main issue with the unblurred data is that the center noise totally explodes, which you can see in your image as well (though not in the particular lineout you've taken).

**Comment by stggh** [comment]
> main issue with the unblurred data is that the center noise totally explodes

The noise gradually accumulates. The centre-line explosion for your data may be some other artefact. Here is the O<sub>2</sub><sup>-</sup> (`O2-ANU1024.txt`) photoelectron `basex` inverse Abel transform, summed (to reduce the noise) centre slice +- 12 rows:
![o2reg](https://user-images.githubusercontent.com/10932229/48170205-8a2ba000-e34a-11e8-85a8-380f859fdb01.png)
Makes your data signal/noise look good! Note, the noise gradually accumulates, toward the image centre, and @DanHickstein is right, the `basex` `reg` parameter does improve the appearance, at the expense of spectral resolution.

**Comment by stggh** [comment]
> better option than simply blurring the image

I agree. Process the raw image, and apply any smoothing in the transform, or later.

I am not 100% convinced that the central transform peak is noise driven, since noise accumulates gradually. Some intensity variation of the raw image, near the centre-line, causes the inverse Abel transform to kick back up.

![image](https://user-images.githubusercontent.com/10932229/48235790-684a2000-e413-11e8-80e4-1e046063e5f8.png)

However, it is likely that I do not know what I am talking about, with respect to plasma discharges. ;-)

**Comment by lemmatum** [comment]
Yea, there is certainly some structure in the center - it's quite apparent in the 2D image.
BTW, it's a laser-plasma experiment, not a discharge. ;)

Do you guys know if there is a way to Abel invert data where the data around the centerline is missing? The target and dense plasma occlude part of my image near the axis of symmetry on some images.

**Comment by stggh** [comment]
Somewhat surprisingly, and in particular, for the `hansenlaw` method(*), the inverse Abel transform starts at the outer (right) half-image edge, iterating along each row toward the image centreline. Each row is independent of the other. So, provided the data is only missing on the inner side of each row, there is no problem. 

NB the transform in the region of missing data is invalid, and may just as well be set to zero (after the transform). 

Hence, apply the transform to the half image, and the set the bad/missing data regions of the transformed image to zero.

(*) I am pretty sure the other non-basex methods `three_point` etc will also be fine. I am not sure about  `basex`.

**Comment by oliverhaas** [comment]
Some thoughts on the topics discussed here:

-----

What PyAbel implicitly assumes as of now is that the data can be zero padded by however many zeros the method uses needs. The basex method is actually the most extreme or worst case since the underlying Gaussian basis doesn't have -- mathematically strictly speaking -- finite support. The resulting error depends on how for this zero padding is wrong and lead to arbitrary large errors, especially at the outer boundaries of the images.
As I have posted in another issue I think a truncation of the integration would be cleaner to do, i.e. for the forward transform 

![missing](https://user-images.githubusercontent.com/10932229/48446030-41f9fb00-e7ec-11e8-9f56-dfb36abb8a24.png) .

For some methods this is equivalent with the current definition (e.g. onion peeling), for other methods it might be somewhat difficult to implement. One still has a "truncation error" of course, but this error is bounded by an integral over the truncated/missing data, which is usually small.

-----

The problem with the your noise in the data set @lemmatum is that the oscillations doesn't seem to damp near the center in your measurement data, and the inverse transform always amplifies oscillations/noise especially near the center due to taking a numerical derivative. If your measurement 
data looked a little bit less oscillatory, more damped, and flat near the center results would be much better.
Physically in your experiment it's reasonable to have more "turbulence" at the center as far as I understand you, but mathematically there is nothing you can do about it except for filtering.
